zkevm-circuits/src/evm_circuit/execution/error_oog_pure_memory.rs

```diff
@@ -0,0 +1,154 @@
+use crate::{
+    evm_circuit::{
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        param::MAX_MEMORY_SIZE_IN_BYTES,
+        step::ExecutionState,
+        util::{
+            constraint_builder::ConstraintBuilder,
+            math_gadget::{IsEqualGadget, IsZeroGadget, LtGadget},
+            memory_gadget::{address_high, address_low, MemoryExpansionGadget},
+            Cell, Word,
+        },
+    },
+    util::Expr,
+};
+use bus_mapping::{
+    eth_types::ToLittleEndian,
+    evm::{GasCost, OpcodeId},
+};
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ErrorOOGPureMemoryGadget<F> {
+    opcode: Cell<F>,
+    address: Word<F>,
+    address_in_range: IsZeroGadget<F>,
+    memory_expansion:
+        MemoryExpansionGadget<F, { MAX_MEMORY_SIZE_IN_BYTES * 2 - 1 }>,
+    insufficient_gas: LtGadget<F, { MAX_MEMORY_SIZE_IN_BYTES * 2 - 1 }>,
+    is_mstore8: IsEqualGadget<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for ErrorOOGPureMemoryGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasPureMemory";
+
+    const EXECUTION_STATE: ExecutionState =
+        ExecutionState::ErrorOutOfGasPureMemory;
+
+    // Support other OOG due to pure memory including CREATE, RETURN and REVERT
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // Query address by a full word
+        let address = cb.query_word();
+
+        // Check if this is an MSTORE8
+        let is_mstore8 = IsEqualGadget::construct(
+            cb,
+            opcode.expr(),
+            OpcodeId::MSTORE8.expr(),
+        );
+        let is_not_mstore8 = 1.expr() - is_mstore8.expr();
+
+        // Get the next memory size and the gas cost for this memory access
+        let memory_expansion = MemoryExpansionGadget::construct(
+            cb,
+            cb.curr.state.memory_size.expr(),
+            address_low::expr(&address)
+                + 1.expr()
+                + (is_not_mstore8 * 31.expr()),
+        );
+
+        // Check if the memory address is too large
+        let address_in_range =
+            IsZeroGadget::construct(cb, address_high::expr(&address));
+        // Check if the amount of gas available is less than the amount of gas
+        // required
+        let (_, memory_gas_cost) = memory_expansion.expr();
+        let insufficient_gas = LtGadget::construct(
+            cb,
+            cb.curr.state.gas_left.expr(),
+            GasCost::FASTEST.expr() + memory_gas_cost,
```

Comment:
I don't understand here why need a extra `GasCost::FASTEST`?

Reply:
That is the constant gas cost of `MLOAD/MSTORE/MSTORE8`, the name of `memory_gas_cost` seem confusing, `memory_expansion_gas_cost` might be better. Fixed in [`9cc8bc7`](https://github.com/appliedzkp/zkevm-circuits/pull/196/commits/9cc8bc77726149d1b6abb225c11c409bbe1d54d8).

(In successful case, the constant gas is added in `SameContextGadget`)

Reply:
Cool!